admin/acceptance.py

```diff
@@ -528,19 +577,196 @@ def start_cluster(self, reactor):
+        d.addCallback(lambda _: perform(make_dispatcher(reactor), commands))
+        return d
+
+    def _create_and_provision(self, reactor, name, retries=10):
+        """
+        Create a node and make it ready for Flocker installation
+        by provisioning it with required software and configuration.
+        The process is transparently retried if necessary.
+
+        :param reactor: The reactor.
+        :param str name: The name.
+        :param int retries: Maximum number of retries.
+        :return: Deferred that fires with the created node when
+            the provisioning is completed.
+        """
+        def create_attempt():
+            d = deferToThreadPool(
+                reactor,
+                reactor.getThreadPool(),
+                self._create_node,
+                name,
             )
+
+            def provision(node):
+                d = self._provision_node(reactor, node)
+
+                def error(failure):
+                    # Destroy a node if we failed to provision it.
+                    self._destroy_node(node)
+                    return failure
+
+                d.addErrback(error)
+                # Make sure too return the node.
+                d.addCallback(lambda _: node)
+                return d
+
+            d.addCallback(provision)
+
+            # Log and discard a failure to keep looping.
+            d.addErrback(write_failure)
+            return d
+
+        d = loop_until(reactor, create_attempt, repeat(0, retries))
+
+        def error(failure):
+            print "Failed to provision node {}".format(name)
+            return failure
+
+        d.addErrback(error)
+        return d
+
+    def _create_nodes(self, reactor, names):
+        """
+        Create and provision a number of nodes with the given names.
+
+        :param reactor: The reactor.
+        :param names: The names.
+        :type names: list of str
+        :return: List of Deferreds each firing when the corresponding node
+            is created and provisioned.
+        """
+        reactor.suggestThreadPoolSize(len(names))
+        return [
+            self._create_and_provision(reactor, name) for name in names
+        ]
+
+    def _make_node_name(self, tag, index):
+        """
+        Construct a node name out of various components.
+        """
+        return "%s-%s-%s-%d" % (
+            self.identity.prefix, self.creator, tag, index,
+        )
+
+    def _add_node_to_cluster(self, reactor, cluster, node, index):
+        """
+        Configure the given node as an agent node for the given cluster.
+
+        :param reactor: The reactor.
+        :param reactor: The cluster.
+        :param Node node: The node.
+        :param int index: The index of the ndoe in the cluster.
+        :return: Deferred that fires with the :param:`node` when it is
+            configured.
+        """
+        node_certnkey = cluster.certificates.add_node(index)
```

Comment:
What does 'certnkey' mean? Is it an abbreviation of something? Should I read it as 'cert_n_key', 'certn_k